H4L updates

YujiLee301 opened the pull request.

YujiLee301 committed:
H4L updates

ram1123 commented on post_proc.py:
```diff
@@ -1,70 +1,21 @@
+import os,sys
+
+from PhysicsTools.NanoAODTools.postprocessing.framework.postprocessor import PostProcessor
+from HZZAnalysisModule import *
```
I think the file `HZZAnalysisModule.py` should be in the PR. No?

YujiLee301 committed:
“Fsr Recovery & Rochester correction”

ram1123 commented on H4LCppModule.py:
```diff
@@ -19,6 +19,17 @@ def __init__(self):
                 ROOT.gROOT.ProcessLine(
                     ".L %s/interface/H4LTools.h" % base)
+        if "/RoccoR_cc.so" not in ROOT.gSystem.GetLibraries():
+            base = "/afs/cern.ch/user/y/yujil/Run3H4l/CMSSW_10_6_20/src/PhysicsTools/NanoAODTools/python/postprocessing/analysis/nanoAOD_vvVBS" 
```
Don't add the absolute path. Add the relative path. Use the environment variable `$CMSSW_BASE`

ram1123 commented on H4LCppModule.py:
```diff
@@ -19,6 +19,17 @@ def __init__(self):
+                    ".L %s/src/RoccoR.cc+O" % base)
+            else:
+                base = "%s/src/PhysicsTools/NanoAODTools" % os.getenv(
+                    "CMSSW_BASE")
```
Similar to this one. In the above comment.

ram1123 commented on H4LCppModule.py:
```diff
@@ -0,0 +1,181 @@
+    def __init__(self):
+        if "/H4LTools_cc.so" not in ROOT.gSystem.GetLibraries():
+            print("Load C++ module")
+            base = "/afs/cern.ch/user/y/yujil/Run3H4l/CMSSW_10_6_20/src/PhysicsTools/NanoAODTools/python/postprocessing/analysis/nanoAOD_vvVBS" 
```
Don't add the absolute path. Add the relative path. Use the environment variable `$CMSSW_BASE`

YujiLee301 committed:
Jet Selection Updated

YujiLee301 committed:
Mela loaded

ram1123 commented on interface/helloworld.h:
Do we need this fine?

ram1123 commented on src/helloworld.cc:
Do we need this file?

ram1123 commented on src/NJettiness.cc:
Fix this?

either remove or fix at least spelling mistake.

also if this creates the issue with compilation then remove it.

ram1123 reviewed (changes requested):
Also update the instructions in README file for compilation and example  syntax using which we can run and check.

YujiLee301 committed:
Delete helloworld.h

YujiLee301 committed:
Delete NJettiness.h

YujiLee301 committed:
Delete NJettiness.cc

YujiLee301 committed:
Delete helloworld.cc

YujiLee301 committed:
Delete proxy.sh

YujiLee301 committed:
Delete condor_setup_lpc.py

YujiLee301 committed:
Delete JetSFMaker.py

YujiLee301 committed:
Delete HZZAnalysisModule.py

YujiLee301 committed:
Delete condor_setup_lxplus.py

YujiLee301 committed:
Add files via upload

YujiLee301 committed:
Remove redundancy

YujiLee301 committed:
Update README.md

ram1123 commented on src/H4LTools.cc:
```diff
@@ -0,0 +1,581 @@
+    
+      if(overlaptag==0){
+        if(((*Jet_pt)[i]>30)&&(fabs((*Jet_eta)[i])<4.7)){
+            std::cout<<"jetindex: "<<i<<"jetID "<<(*Jet_jetId)[i]<<" puID "<<(*Jet_puId)[i]<<std::endl;
```
Comment this line or add the verbosity level so that we can switch on when we need to print this line.

ram1123 commented on interface/H4LTools.h:
```diff
@@ -0,0 +1,176 @@
+};
+
+H4LTools::H4LTools(){
+  std::string DATAPATH = "/afs/cern.ch/user/y/yujil/Run3H4l/CMSSW_10_6_20/src/PhysicsTools/NanoAODTools/python/postprocessing/analysis/nanoAOD_vvVBS";
```
Put the relative path instead of the absolute path.

This is pointing to your afs path, and it won't work for others.

ram1123 replied:
One of easiest way that I can think of is:

Just replace:

```cpp
  std::string DATAPATH = "/afs/cern.ch/user/y/yujil/Run3H4l/CMSSW_10_6_20/src/PhysicsTools/NanoAODTools/python/postprocessing/analysis/nanoAOD_vvVBS";
      DATAPATH += "/KalmanMuonCalibrationsProducer/data/roccor.Run2.v5/RoccoR2018UL.txt";

```

with 

```cpp
  std::string DATAPATH = "";
      DATAPATH += "KalmanMuonCalibrationsProducer/data/roccor.Run2.v5/RoccoR2018UL.txt";

```

and run the code `python post_proc.py` from path:

```bash
/afs/cern.ch/user/y/yujil/Run3H4l/CMSSW_10_6_20/src/PhysicsTools/NanoAODTools/python/postprocessing/analysis/nanoAOD_vvVBS
```

ram1123 reviewed (changes requested):
- Missing file `keep_and_drop.txt`
- Missing directory and files inside it: `KalmanMuonCalibrationsProducer/`

YujiLee301 committed:
Add files via upload

YujiLee301 committed:
Update README.md

YujiLee301 committed:
Update README.md

YujiLee301 committed:
Update README.md

YujiLee301 committed:
RoccoR Added

YujiLee301 committed:
Path fixed

ram1123 commented on README.md:
Replace 

`git@github.com:osWW-VBS/nanoAOD_skim.git`

with 

`git@github.com:YujiLee301/nanoAOD_vvVBS.git`


Also, I can see that in instruction the name that is set for the directory is `nanoAOD_skim` then why you have directory named `nanoAOD_vvVBS`?

Basically the git clone line will be:

```bash
git clone git@github.com:YujiLee301/nanoAOD_vvVBS.git PhysicsTools/NanoAODTools/python/postprocessing/analysis/nanoAOD_skim
```

ram1123 replied:
Replace all the instance of `nanoAOD_vvVBS` to `nanoAOD_skim`.

ALso, from some of your .py and .h/.cc files, where you added the path. And from this README file.

ram1123 replied:
Also, to generalize the name I update the repository name to `nanoAOD_skim`. :)

ram1123 commented on KalmanMuonCalibrationsProducer/python/__init__.pyc:
never commit any executable files. Like `.pyc` file. Delete this  file.

ram1123 commented on KalmanMuonCalibrationsProducer/python/__init__.py:
Do we need this file? You created this or it automatically created?

If automatically then delete this.

ram1123 commented on KalmanMuonCalibrationsProducer/data/roccor.Run2.v5/RoccoR.cc:
I think we don't need this .cc and .h file. YOu already added them in src and interface directory. So, you can delete them from here.

ram1123 commented on JetSFMaker.py:
Why you deleted this file?

ram1123 commented on condor_setup_lxplus.py:
Why deleted this file?

ram1123 commented on proxy.sh:
don't delete this file.

ram1123 commented on post_proc.py:
```diff
@@ -1,70 +1,27 @@
-# testfile = "file:/tmp/rasharma/11F21FB5-7535-B249-8F59-1173DC384F33.root"
-
-entriesToRun = 1000  # 0 if need to run over all entries else put number of entries to run
-isMCTrueFalse = True
```
I will keep line 14-65. May be commented for now. Because we will need them later.

YujiLee301 committed:
Add files via upload

YujiLee301 committed:
Add files via upload

YujiLee301 committed:
Delete __init__.pyc

YujiLee301 committed:
Delete RoccoR.cc

YujiLee301 committed:
Delete RoccoR.h

YujiLee301 committed:
Update README.md

YujiLee301 committed:
Update H4LCppModule.py

ram1123 commented on README.md:
Replace

```
git@github.com:osWW-VBS/nanoAOD_skim.git
```

with

```
git@github.com:YujiLee301/nanoAOD_skim.git
```

on line 26

ram1123 commented:
Its working now. So, no further comments from me. Thank you @YujiLee301

ram1123 commented:
General question:

I see two functions named `ZZSelection`.

One defined in  https://github.com/ram1123/nanoAOD_skim/blob/1ca32494774555ac82875a5079a99486e4ed0590/Utils.py#L4


 and another defined in https://github.com/ram1123/nanoAOD_skim/blob/1ca32494774555ac82875a5079a99486e4ed0590/src/H4LTools.cc#L228

Which one is used in the code?

ram1123 commented on interface/H4LTools.h:
```diff
@@ -0,0 +1,176 @@
+      TTreeReaderArray<float> *Jet_btagDeepC = nullptr;
+      TTreeReaderArray<int> *Jet_jetId = nullptr;
+      TTreeReaderArray<int> *Jet_puId = nullptr;
+
```
You can remove many blank lines

ram1123 committed:
Update keep_and_drop.txt

ram1123 committed:
Update README.md

Following updates:

- Updated to `CMSSW_10_6_30` from `CMSSW_10_6_20`
- Updated to the latest version of official nanoAOD-tools
- Added back the instruction of batch job.

ram1123 commented:
@YujiLee301 FYI: I updated the `keep_and_drop.txt` and the README files.

In the README file, I updated two more pieces of information that you should update.

1. Updated to the latest recommended CMSSW. I already tested it, and it's working.
2. Updated to the latest nanoAOD-tools. Again, I tried, and it's working.

ram1123 committed:
Update README.md

YujiLee301 committed:
MELA updates

YujiLee301 commented:
> General question (@YujiLee301 ):
> 
> I see two functions named `ZZSelection`.
> 
> One defined in
> 
> https://github.com/ram1123/nanoAOD_skim/blob/1ca32494774555ac82875a5079a99486e4ed0590/Utils.py#L4
> 
> and another defined in
> 
> https://github.com/ram1123/nanoAOD_skim/blob/1ca32494774555ac82875a5079a99486e4ed0590/src/H4LTools.cc#L228
> 
> Which one is used in the code?

The second one is used in Cpp Module. And the first one is in python versiom.

ram1123 commented on KalmanMuonCalibrationsProducer/interface/RoccoR.h:
This file (and corresponding .cc file) is at two places. Which one is used?

Remove the one that we are not using.

YujiLee301 committed:
Delete KalmanMuonCalibrationsProducer/interface directory

YujiLee301 committed:
Delete KalmanMuonCalibrationsProducer/plugins directory

YujiLee301 committed:
Delete KalmanMuonCalibrationsProducer/python directory

YujiLee301 committed:
Delete KalmanMuonCalibrationsProducer/src directory

YujiLee301 committed:
Delete BuildFile.xml

YujiLee301 committed:
Trigger selection added

ram1123 commented on Helper.py:
```diff
@@ -1,16 +1,16 @@
 import ROOT
 
 PI=3.14159
```
In this function you should also pass the year as the list of triggers. As the list of triggers depends on year.

also instead of many if else add all triggers in the array and then use it. @YujiLee301

YujiLee301 committed:
Trigger Updated

YujiLee301 committed:
Update keep_and_drop.txt with Mela branches

ram1123 committed:
Update keep_and_drop.txt

ram1123 reviewed (changes requested):
One important piece of information that I just remembered.

@YujiLee301: You corrected leptons for the FSR. We need to save the leptons after including the FSR information. When we added leptons in `keep_and_drop` text file, it only added leptons before correction (i.e. which is already there in original nanoAOD root file).

ram1123 commented:
Hi @YujiLee301 ,

There are two issues with the code:

1. All the default branches coming from the original nanoAOD root files are empty. Please check and fix it.
2. Also, add the jets information that are cleaned from the electrons.

Let me know if you have any confusion.

ram1123 commented:
> * All the default branches coming from the original nanoAOD root files are empty [1]. Please check and fix it.

So, this is not true. I check the information on low statistics with the wrong MC sample. If I check the other branches except the FatJet, I can see that the information is there.

YujiLee301 committed:
Update keep_and_drop.txt

YujiLee301 committed:
Sync for Lepton

ram1123 commented:
Hi @YujiLee301 ,

As we chatted earlier, we plan to generalize this for 4l, 2l2q and 2l2nu.

The function that you made, `ZZSelection()`, does everything.

But it would be great if we add two more function say `Z1Selection()` and `Z2Selection()`. And then you call these two functions inside `ZZSelection()`.

If you do like this then we could make and call separate `Z2Selections()` depending on the number of leptons inside `ZZSelection()`. 

Let me know if you have confusion. We can talk.

YujiLee301 committed:
Add Z selection

ram1123 reviewed (commented):
Hi @YujiLee301 ,

Did you change something how we select Z1 and Z2?

As the pT value is different now.

# New

<img width="1007" alt="new" src="https://github.com/ram1123/nanoAOD_skim/assets/5220316/cb0a15a1-3360-4716-b788-a8954dd4e839">

# old
<img width="1006" alt="old" src="https://github.com/ram1123/nanoAOD_skim/assets/5220316/3117cb93-196c-417c-95d3-cff6c071746d">

YujiLee301 committed:
Lepton Module Changed

ram1123 commented on H4LCppModule.py:
```diff
@@ -199,12 +199,15 @@ def analyze(self, event):
         nZXCRFailedLeptons=0
         isMC = True
-        passedTrig = PassTrig(event, 2017)
+        passedTrig = PassTrig(event, 2018)
```
don't hard code this here. 

Try to pass the year from here: https://github.com/ram1123/nanoAOD_skim/blob/7b26df21bb996afee7b47d960462e894eff94f9a/post_proc.py

ram1123 commented on interface/H4LTools.h:
```diff
@@ -248,7 +269,7 @@ class H4LTools {
 H4LTools::H4LTools(){
   std::string DATAPATH = "";
-  DATAPATH += "KalmanMuonCalibrationsProducer/data/roccor.Run2.v5/RoccoR2017UL.txt";
+  DATAPATH += "KalmanMuonCalibrationsProducer/data/roccor.Run2.v5/RoccoR2018UL.txt";
```
don't hard code this here. 

Try to pass the year from here: https://github.com/ram1123/nanoAOD_skim/blob/7b26df21bb996afee7b47d960462e894eff94f9a/post_proc.py

Then use the path based on year.

If (year = 2017) {
DATAPATH += "KalmanMuonCalibrationsProducer/data/roccor.Run2.v5/RoccoR2017UL.txt";
}
else If (year = 2018) {
DATAPATH += "KalmanMuonCalibrationsProducer/data/roccor.Run2.v5/RoccoR2018UL.txt";
}
else  {
 ERROR
}

ram1123 reviewed (commented):
Hi @YujiLee301 ,

It seems fine. I will try to run and check this later.

Just added one general comment. That we need to propagate year from post_proc.py file to everywhere.

You can get year automatically in post_proc something like this:

https://github.com/ram1123/nanoAOD_skim/blob/ee7343a7170a1f3ead3476e732a35ee11949abad/post_proc.py#L15-L29

YujiLee301 committed:
hard code removed

YujiLee301 committed:
Reading branch modified

ram1123 commented on src/H4LTools.cc:
```diff
@@ -0,0 +1,855 @@
+        int overlaptag=0;
+        TLorentzVector jettest;
+        jettest.SetPtEtaPhiM(Jet_pt[i],Jet_eta[i],Jet_phi[i],Jet_mass[i]);
+        for(unsigned int ie=0;ie<ele.size();ie++){
```
@YujiLee301: You should clean with only those jets that passed your pre-selection. I mean the pT and eta cut should be applied before checking the deltaR with leptons.

YujiLee301 committed:
Jet Selection fixed

ram1123 commented on Helper.py:
```diff
@@ -0,0 +1,146 @@
+    
+    TriggerList = []
+    PassTrig = False
+    if (year == 2018):
```
@YujiLee301 : For completeness, add all three-year trigger list in **one separate file.** Then import it here.

This separate file should be a [YAML file](https://www.redhat.com/en/topics/automation/what-is-yaml#). One file for each year. Based on the year it will call the appropriate YAML file. And this file supposes to contain all input information that the code will need—for example, Trigger list, Rochester correction file, hard-coded values, etc.

YujiLee301 committed:
tight Electron corrected

ram1123 commented on Input_2017.yml:
Apart from the trigger and Rochester correction files, it would be good if you can also add the following information in the YAML file

- Lepton pT, eta
- Lepton ID
- Jet pT, eta
- Mll cut
- mZ cut, etc

ram1123 commented:
Comparison for the latest version of nanoAOD and miniAOD:

![image_2023_08_15T10_04_54_915Z](https://github.com/ram1123/nanoAOD_skim/assets/5220316/7061bbde-0c3b-4425-beb0-75586995de09)

ram1123 commented:
As we have quite a good comparison between the nanoAOD and miniAOD. So, I am going to merge this.

Some minor improvements like:

- Adding other parameters to the YAML file
- More detailed cut-flow chart: As a histogram in the output root file, it will always remain stuck to the output root file.
- Some improvements will come with a detailed comparison of miniAOD and nanoAOD.
